### PR TITLE
ARM FPSCR bit corrections

### PR DESCRIPTION
**Pull request by smolt**
Correct IeeeFlags exception masks and FloatingPointControl RoundingModes for ARM fpscr

These changes will help GDC @ibuclaw and LDC @klickverbot with their 32-bit arm targets.  Tested on cortex-a9 armv7.

### CODE REVIEW

File: std/math.d [hunk 1]
```diff
@@ -4066,11 +4066,11 @@ private:
...
-            INEXACT_MASK   = 0x1000,
-            UNDERFLOW_MASK = 0x0800,
-            OVERFLOW_MASK  = 0x0400,
-            DIVBYZERO_MASK = 0x0200,
-            INVALID_MASK   = 0x0100
+            INEXACT_MASK   = 0x10,
+            UNDERFLOW_MASK = 0x08,
+            OVERFLOW_MASK  = 0x04,
+            DIVBYZERO_MASK = 0x02,
+            INVALID_MASK   = 0x01
```
Comment (ibuclaw) [inline std/math.d]: You see, what you are doing here is setting them to the bits representing exceptions.  What the name implies is the mask.

This change is wrong.

Comment (ibuclaw) [inline std/math.d]: What should be fixed is ``core.stdc.fenv`` - which is not cross-platform friendly at all @MartinNowak  :-)

Comment (MartinNowak) [inline std/math.d]: Yes, seems like those constants should be split up into architectures.
https://github.com/D-Programming-Language/druntime/blob/e66f03494a13c7ce13bbe4c260ce689682af865e/src/core/stdc/fenv.d#L250

Comment (MartinNowak) [inline std/math.d]: Seems like the change is correct though, as the docs say `IEEE exception status flags ('sticky bits')` and `These flags indicate that an exceptional floating-point condition has occurred`.
Someone had the bad idea to append this mask suffix.

**Comment by MartinNowak** [comment]
You also have ask to read the flag?

**Comment by smolt** [comment]
@ibuclaw - it does seem incorrect to have these defined here, just fixing current content.

@MartinNowak - the asm to read and set ARM fpscr is already in both LDC and GDC, just the constants were wrong.  Actually the ARM asm is slightly wrong for `resetIeeeFlags()` in LDC and I think GDC too because it clears entire fpscr, which includes rounding modes or other ARM flags (default-nan, flush-to-zero, etc).  I am going to make separate LDC pull for that. @ibuclaw - do you want same for GDC?

The current code does this:
```D
uint old = getIeeeFlags();
old &= ~0b11111; // http://infocenter.arm.com/help/topic/com.arm.doc.ddi0408i/Chdfifdc.html
__asm("vmsr FPSCR, $0", "r", old);
```

but `getIeeeFlags()` has already masked off the flags, so `old` doesn't have other bits set like rounding mode.  Better, and simpler  - using LDC style-inline:
```D
cast(void) __asm!uint
          ("vmrs $0, fpscr;"
           "bic $0, #0x1f;"
           "vmsr fpscr, $0", "=r");
```

**Comment by jpf91** [comment]
> do you want same for GDC?

Yes, please open a pull request for GDC as well.

**Comment by MartinNowak** [comment]
Why not maintaining such gdc/ldc specific snippets in phobos itself? It seems pretty cumbersome to track down such changes and port them.

**Comment by ibuclaw** [comment]
I think last time it was discussed, it was agreed that having a mixture of ``asm { ... }`` /  ``__asm(...)`` / ``asm { "..." }`` styles was not a good idea.  I'm sure DMD maintainers still wish to ignore the fact that three competing styles exist for inline assembly.  Or if they aren't ignoring it, they don't want to acknowledge it by allowing it in mainstream libraries.

**Pull request merged**